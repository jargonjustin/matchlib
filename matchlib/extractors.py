import new

def extractor(func):
    "Decorator to convert a function into an extractor object"
    return new.classobj(func.__name__, (object,), {
        '__match__': staticmethod(func)
    })

def caseclass(*attrs):
    """Creates a caseclass useful as an algebraic data type
    
    A caseclass handles the common case of a class that is defined primarily by
    it's structure (that is, the attribute's it contains as data.)  Abstract
    syntax trees and simple algebraic data-types are common examples.  Caseclasses
    contain a constructor taking it's arguments in the order they were given,
    methods for matching, equality comparison, and displaying themselves.
    
    class Num(caseclass('value')):
        def __init__(self, value):
            self.value = int(value)
    
    Num(3) == Num(3)    # => True
    repr(Num('5'))      # => Num(5)
    
    """
    return new.classobj('caseclass', (__CaseClass,), {
        '__slots__': attrs
    })

def matchattrs(*attrs):
    """Generates a classmethod to be used as __match__ in an extractor.
    
    The generated method checks that the object is an instance of the class to
    which the __match__ method is attached.  It then extracts the listed
    listed attributes.
    
    class Add(object):
        def __init__(self, lhs, rhs):
            self.lhs, self.rhs = lhs, rhs
        __match__ = matchattrs('lhs', 'rhs')
    
    match Add(1, 2) with:
        case Add(x, y): print x, y
    
    """
    def __match__(cls, obj):
        if isinstance(obj, cls):
            return tuple([ getattr(obj, attr) for attr in attrs ])
    return classmethod(__match__)

class __CaseClass(object):
    __slots__ = ()
    
    def __init__(self, *args):
        if len(args) != len(self.__slots__):
            raise TypeError, "%s.__init__() takes exactly %d arguments (%d given)" % (self.__class__.__name__, len(self.__slots__), len(args))
        for (attr, value) in zip(self.__slots__, args):
            setattr(self, attr, value)
    
    @classmethod
    def __match__(cls, obj):
        if isinstance(obj, cls):
            return tuple([ getattr(obj, attr) for attr in cls.__slots__ ])
    
    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            return False
        for attr in self.__slots__:
            if not getattr(self, attr) == getattr(other, attr):
                return False
        return True
    
    def __ne__(self, other):
        return not self == other
    
    def __hash__(self):
        hashed = []
        for attr in self.__slots__:
            try:
                hashed.append(hash(getattr(self, attr)))
            except TypeError:
                continue
        return hash(tuple(hashed))
    
    def __repr__(self):
        return "%s(%s)" % (self.__class__.__name__, ", ".join([ repr(getattr(self, attr)) for attr in self.__slots__ ]))
    

