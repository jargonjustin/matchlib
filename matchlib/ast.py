from matchlib.extractors import caseclass
import token

class PVar(caseclass('name')): pass
class PSym(caseclass('name')): pass

class PLit(caseclass('token')):
    def __repr__ (self):
        return 'PLit(%s, %s)' % (token.tok_name[self.token[0]], repr(self.token[1]))

class PTuple(caseclass('patterns')): pass
class PList(caseclass('patterns')): pass
class PDict(caseclass('pairings')): pass

class PExtractor(caseclass('extractor', 'patterns')): pass

class PBind(caseclass('name', 'pattern')): pass
class PSplat(caseclass('name')): pass
