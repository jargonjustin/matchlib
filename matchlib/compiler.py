import os, imp, tempfile, time, token, tokenize

import matchlib
from matchlib.ast import *
from matchlib import pattern_parser as parser
from matchlib import codegen as g
from matchlib import debug

### Compiler Frontend and Public API ###

def main (argv=None):
    "Runs the compiler frontend and exits"
    
    from optparse import OptionParser
    import os, sys, traceback
    
    if argv is None:
        argv = sys.argv
    
    def fail (message):
        print >>sys.stderr, '%s: %s' % (os.path.basename (sys.argv[0]), message)
        return 1
    
    def fail_exception (exctype, exception):
        traceback.print_exception (exctype, exception, tb=None, file=sys.stderr)
        return 1
    
    optparser = OptionParser (usage="%prog [options] file...")
    optparser.add_option ('-o', dest='outFile', help='output filename', metavar="<file>")
    (options, args) = optparser.parse_args ()
    
    if not args:
        return fail ('no input files')
    elif options.outFile and len (args) != 1:
        return fail ('output specified for multiple inputs')
    
    try:
        if options.outFile is None:
            for infile in args:
                (path, ext) = os.path.splitext (infile)
                if ext != '.mpy':
                    return fail ('invalid file type (expected .mpy, got %s)' % ext)
                compile_file (infile, path + '.py')
        else:
            compile_file (args[0], options.outFile)
    except IOError, err:
        return fail ('%s: %s' % (err.strerror, err.filename))
    except IndentationError, err:
        return fail_exception (IndentationError, err)
    except SyntaxError, err:
        return fail_exception (SyntaxError, err)
    except parser.ParseError, err:
        parser.print_failure (err)
        return 1
    else:
        return 0

def compile_as_module (modulename, filename):
    "Compiles and returns the filename as a module"
    inp, out = None, None
    try:
        inp = open (filename, 'U')
        
        # Create and open a temporary file for writting.  We need to do this
        # because imp.load_module requires a file object (a StringIO won't)
        # pass muster
        (fd, outname) = tempfile.mkstemp (suffix='.py')
        out = os.fdopen (fd, 'w+')
        
        # Compile from one file to the other
        compile_piped (inp, out, filename)
        
        # Seek to the begining of the compiled source
        out.seek (0)
        
        # TODO: the compiled output goes in whatever is passed as the filename
        # parameter + 'c', should we leave it as the .mpy file or switch the
        # extension to '.py' to generate a normally loadable '.pyc'?
        return imp.load_source (modulename, filename, out)
    finally:
        # Close the files and remove the temporary file
        if inp: inp.close ()
        if out:
            out.close ()
            os.remove (outname)

def compile_file (input_filename, output_filename):
    "Compiles the input file and writes the output to the output file"
    # Opens the files and passes them off to compile_piped, cleaning up
    # afterwards if necessary
    inp, out = None, None
    try:
        inp = open (input_filename, 'U')
        out = open (output_filename, 'w')
        compile_piped (inp, out, input_filename)
    finally:
        if inp: inp.close ()
        if out: out.close ()

def compile_piped (inp, out, filename=None):
    
    """Reads the source from one file-like object and writes the compiled output to another
    
    The optional filename argument is used for debugging annotations.
    
    May throw a matchlib.pattern_parser.ParseError, IndentationError, or SyntaxError if
    there is a syntactic error in the input.
    """
    
    # Parse and compile the input
    compiled_lines = compile_lines (parser.parse_handle (inp, filename))
    # Annotate and serialize the compiled output
    compiled_source = ''.join (parser.unlines (debug.annotate (compiled_lines, filename)))
    # Raise an exception if there is a syntactic error
    debug.check_syntax (compiled_source)
    # Write out the debugging header
    out.write ("# compiled by matchlib version %s on %s\n" % (matchlib.__version__, time.asctime ()))
    if filename is not None:
        out.write ("# original source file: %s\n" % filename)
    out.write ("\n")
    out.write ("# DO NOT MODIFY THIS FILE\n")
    out.write ("# This file is automatically generated, any modifications will likely be overwritten\n\n")
    # Write out the source
    out.write (compiled_source)

### Compiler Internals ###
# Here there be dragons

def compile_lines (lines):
    
    """Transforms lines of the form generated by pattern_parser.parse and
       pattern_parser.lines into lines without match statements, replacing
       them with equivilent python code."""
    
    lines = iter (lines)
    while True:
        line = lines.next ()
        if line[0] is parser.MATCH:
            compiler = MatchCompiler (line[1])
            while True:
                line = lines.next ()
                if line[0] is parser.CASE:
                    (patterns, predicate) = line[1]
                    body = list (compile_lines (parser.extract_block (lines)))
                    compiler.add_case (patterns, predicate, body)
                elif line[0] is parser.END:
                    break
                else:
                    parser.fail ('expected case statement', line[1][0])
            (funcLines, actionLines) = compiler.compile ()
            for line in funcLines: yield line
            for line in actionLines: yield line
        else:
            yield line

class MatchCompiler (object):
    
    def __init__ (self, match_expr):
        self.match_expr = match_expr
        self.cases = list ()
    
    def add_case (self, patterns, predicate, body):
        self.cases.append ((patterns, predicate, body))
    
    def compile (self, funcTarget=None, actionTarget=None):
        (funcId, params, variables, funcCode) = self.__compile_function (funcTarget)
        actionCode = self.__compile_actions (funcId, params, variables, actionTarget)
        return (funcCode, actionCode)
    
    def __extractors (self):
        extractors, stack = set (), list ()
        for (patterns, predicate, body) in self.cases:
            stack.extend (patterns)
        while stack:
            pattern = stack.pop ()
            if isinstance (pattern, PExtractor):
                extractors.add (pattern.extractor)
            elif isinstance (pattern, PTuple) or isinstance (pattern, PList):
               stack.extend (pattern.patterns)
            elif isinstance (pattern, PDict):
                stack.extend ([
                    pairing[1] for pairing in pattern.pairings
                    if not isinstance (pairing, PSplat)
                ])
            elif isinstance (pattern, PBind):
                stack.append (pattern.pattern)
        return extractors
    
    def __compile_function (self, target=None):
        funcId = g.Id (g.UniqueName ())
        bindingMap = dict ()
        params = map (g.Id, self.__extractors ())
        gen = g.CodeGenerator (target)
        
        binding = g.Id (g.UniqueName ())
        patternCompiler = PatternCompiler (binding, gen.target)
        
        gen.Def (funcId, binding, *params)
        
        for case_id, (patterns, predicate, body) in enumerate (self.cases):
            for pat_id, pattern in enumerate (patterns):
                indent = patternCompiler.compile (pattern, predicate)
                if len (patternCompiler.bindings) > 0:
                    variables, bindings = zip (*patternCompiler.bindings.items ())
                else:
                    variables, bindings = list (), list ()
                bindingMap[pattern] = variables
                gen.Return (g.Tuple (g.lift (case_id), g.lift (pat_id), *bindings))
                gen.dedent (indent)
        gen.dedent ()
        
        return (funcId, params, bindingMap, gen.target)
    
    def __compile_actions (self, funcId, args, variables, target=None):
        gen = g.CodeGenerator (target)
        binding = g.Id (g.UniqueName ())
        
        gen.Assign (binding, g.Call (funcId, self.match_expr, *args))
        
        gen.If (g.Is (binding, g.lift (None)))
        gen.Raise (g.Id ('ValueError'), g.lift ("pattern match failed"))
        gen.dedent ()
        
        for case_id, (patterns, predicate, body) in enumerate (self.cases):
            gen.ElIf (g.Is (g.Index (binding, g.lift (0)), g.lift (case_id)))
            started = False
            for pat_id, pattern in enumerate (patterns):
                predicate = g.Is (g.Index (binding, g.lift (1)), g.lift (pat_id))
                if started:
                    gen.ElIf (predicate)
                else:
                    gen.If (predicate)
                    started = True
                if len (variables[pattern]) > 0:
                    gen.Assign (g.Tuple (*map (g.Id, variables[pattern])),
                                g.Slice (binding, g.lift (2), None))
                else:
                    gen.line (g.Id ('pass'))
                gen.dedent ()
            gen.inline (body)
        
        return gen.target

class PatternCompiler (object):
    """Generates the code needed to check if an object matches a pattern"""
    
    def __init__ (self, binding, target=None):
        self.gen = g.CodeGenerator (target)
        self.binding = binding
    
    def compile (self, pattern, predicate):
        self.bindings = dict ()
        
        initialIndent = self.gen.indent
        self.__compile_pattern (pattern, self.binding)
        if predicate:
            if len (self.bindings) > 0:
                variables, bindings = zip (*self.bindings.items ())
                self.gen.Assign (g.Tuple (*map (g.Id, variables)), g.Tuple (*bindings))
            self.gen.If (predicate)
        
        return self.gen.indent - initialIndent
    
    def __compile_pattern (self, pattern, binding):
        if isinstance (pattern, PLit):
            self.gen.If (g.Eq (binding, g.Token (pattern.token)))
        elif isinstance (pattern, PSym):
            self.gen.If (g.Eq (binding, g.Id (pattern.name)))
        elif isinstance (pattern, PVar):
            self.__bind (pattern.name, binding)
        elif isinstance (pattern, PTuple):
            self.__emit_type_check ('__builtins__.tuple', binding)
            self.__compile_listlike (pattern.patterns, binding)
        elif isinstance (pattern, PList):
            self.__emit_type_check ('__builtins__.list', binding)
            self.__compile_listlike (pattern.patterns, binding)
        elif isinstance (pattern, PDict):
            self.__emit_type_check ('__builtins__.dict', binding)
            for pairing in pattern.pairings:
                if isinstance (pairing, PSplat):
                    self.__bind (pairing.name, binding)
                else:
                    self.gen.If (g.In (g.Token (pairing[0].token), binding))
                    self.__compile_pattern (pairing[1], g.Index (binding, g.Token (pairing[0].token)))
        elif isinstance (pattern, PExtractor):
            extractedBinding = g.Id (g.UniqueName ())
            self.gen.Assign (extractedBinding,
                             g.Call (g.Method (g.Id (pattern.extractor), '__match__'), binding))
            self.gen.If ( g.IsNot (extractedBinding, g.lift (None)))
            self.__compile_listlike (pattern.patterns, extractedBinding)
        elif isinstance (pattern, PBind):
            self.__bind (pattern.name, binding)
            self.__compile_pattern (pattern.pattern, binding)
        else:
            raise TypeError, 'expected pattern (got %s)' % type(pattern)
    
    def __bind (self, name, binding):
        if name != '_':
            if name in self.bindings:
                self.gen.If (g.Eq (self.bindings[pattern.name], binding))
            else:
                self.bindings[name] = binding
    
    def __emit_type_check (self, typename, binding):
        self.gen.If (g.Call (g.Id ('__builtins__.isinstance'), binding, g.Id(typename)))
    
    def __compile_listlike (self, patterns, binding):
        splatIndex = None
        for index in range (len (patterns)):
            if isinstance (patterns[index], PSplat):
                splatIndex = index
                break
        if splatIndex is None:
            self.gen.If (g.Eq (g.Call (g.Id ('__builtins__.len'), binding), g.lift (len (patterns))))
            for index in range (len (patterns)):
                self.__compile_pattern (patterns[index], g.Index (binding, g.lift (index)))
        else:
            self.gen.If (g.Gt (g.Call (g.Id ('__builtins__.len'), binding), g.lift (len (patterns) - 2)))
            sliceEnd = -(len (patterns) - splatIndex - 1)
            if sliceEnd == 0:
                sliceEnd = None
            else:
                g.lift (sliceEnd)
            self.__bind (patterns[splatIndex].name, g.Slice (binding, g.lift (splatIndex),  sliceEnd))
            for index in range (len (patterns)):
                if index == splatIndex: continue
                elif index > splatIndex:
                    index = len (patterns) - index
                self.__compile_pattern (patterns[index], g.Index (binding, g.lift (index)))
    

if __name__ == '__main__':
    import sys
    sys.exit(main(sys.argv))
