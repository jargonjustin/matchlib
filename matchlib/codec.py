import codecs, encodings
from cStringIO import StringIO

from matchlib.compiler import compile_piped

class MatchPythonStreamReader(codecs.StreamReader):
    def __init__(self, *args, **kwargs):
        super(MatchPythonStreamReader, self).__init__(*args, **kwargs)
        self.__compiled = None
    
    def read(*args, **kwargs):
        return unicode(self.__compile__().read(*args, **kwargs))
    
    def readline(self, *args, **kwargs):
        return unicode(self.__compile__().readline(*args, **kwargs), 'utf8')
        
    def readlines(self, *args, **kwargs):
        return [
            unicode(line, 'utf8') for line in
            self.__compile__().readlines(*args, **kwargs) ]
    
    def __compile__(self):
        if self.__compiled is None:
            self.__compiled = StringIO()
            compile_piped(self.stream, self.__compiled)
            self.__compiled.reset()
        return self.__compiled

def search_function(encoding):
    if encoding != 'match-python':
        return None
    utf8 = encodings.search_function('utf8')
    return codecs.CodecInfo( name='match-python', 
        encode=utf8.encode, decode=utf8.decode,
        incrementalencoder=utf8.incrementalencoder,
        incrementaldecoder=utf8.incrementaldecoder,
        streamreader=MatchPythonStreamReader,
        streamwriter=utf8.streamwriter)

def register():
    codecs.register(search_function)
