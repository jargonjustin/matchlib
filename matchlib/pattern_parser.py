import token, tokenize, sys
from cStringIO import StringIO
from matchlib.ast import *

def parse_file (filename):
    inp = None
    try:
        inp = open (filename)
        return parse_handle (inp, filename)
    finally:
        if inp: inp.close ()

def parse_string (source, filename=None):
    return parse_handle (StringIO (source), filename)

def parse_handle (handle, filename=None):
    return parse (lines (handle.readline), filename)

### Magic Numbers ###
LINE, BLOCK, END = 1, 2, 3
MATCH, CASE = 4, 5

INDENTS = [ BLOCK, MATCH, CASE ]

### Parser ###

def parse (lines, filename=None):
    lines = iter(lines)
    try:
        while True:
            line = lines.next ()
            match_expr = parse_match_statement (line)
            if match_expr is None:
                fail_on_keywords (line[1], 'match', 'with')
                yield line
            else:
                yield (MATCH, match_expr)
                while True:
                    line = lines.next ()
                    if line[0] is END:
                        yield line
                        break
                    yield (CASE, parse_case_statement (line))
                    for line in extract_block (parse (lines)):
                        yield line
    except ParseError, err:
        err.filename = filename
        raise

def parse_match_statement (line):
    if line[0] is BLOCK and \
       len (line[1]) > 2 and \
       iskeyword ('match', line[1][0]) and \
       iskeyword ('with', line[1][-3]):
        return line[1][1:-3]

### Case Statement Parser ###

def parse_case_statement (line):
    if line[0] is not BLOCK:
        fail ("expected case statement")
    elif not iskeyword ('case', line[1][0]):
        fail ("expected case statement", line[1][0])
    elif len (line[1]) < 4:
        fail ("empty case statement", line[1][0])
    expr_iter = PeekableIterator (line[1][1:-2])
    patterns = [ parse_case_pattern (expr_iter) ]
    while True:
        try:
            tok = expr_iter.next ()
        except StopIteration:
            return (patterns, None)
        if iskeyword ('if', tok):
            return (patterns, list (expr_iter))
        elif isop (',', tok):
            patterns.append (parse_case_pattern (expr_iter))
        else:
            fail ('illegal token in case statement', tok)

def parse_case_pattern (expr_iter):
    try:
        pat = parse_nonbinding_pattern (expr_iter)
        if iskeyword ('as', expr_iter.peek ()):
            expr_iter.next ()
            ident = expr_iter.next ()
            if ident[0] is not token.NAME:
                fail ('binding pattern must bind to a name', ident)
            return PBind (ident[1], pat)
        else:
            return pat
    except StopIteration:
        fail ('premature end of pattern expression')

def parse_nonbinding_pattern (expr_iter):
    tok = expr_iter.next ()
    lit = parse_literal_token (tok)
    if lit is not None:
        return lit
    elif tok[0] == token.NAME:
        idents = [ tok[1] ]
        while True:
            tok = expr_iter.peek ()
            if isop ('.', tok):
                expr_iter.next ()
                tok = expr_iter.next ()
                if tok[0] is not token.NAME:
                    fail ('illegal identifier in case expression', tok)
                idents.append (tok[1])
            else:
                break
        tok = expr_iter.peek ()
        if isop ('(', tok):
            expr_iter.next ()
            return PExtractor ('.'.join (idents), parse_listlike (expr_iter, parse_case_pattern, ')'))
        elif len (idents) > 1:
            return PSym ('.'.join (idents))
        else:
            return PVar (idents[0])
    elif isop ('[', tok):
        return PList (parse_listlike (expr_iter, parse_case_pattern, ']'))
    elif isop ('(', tok):
        return PTuple (parse_listlike (expr_iter, parse_case_pattern, ')'))
    elif isop ('{', tok):
        return PDict (parse_listlike (expr_iter, parse_dict_entry, '}', allowsplat=False))
    else:
        fail ('illegal case expression', tok)


def parse_dict_entry (expr_iter):
    key_token = expr_iter.next ()
    key_lit = parse_literal_token (key_token)
    if key_lit is None:
        fail ('dictionary key must be a literal', key_token)
    colon_token = expr_iter.next ()
    if not isop (':', colon_token):
        fail ('dictionary entries must contain a colon', colon_token)
    return (key_lit, parse_case_pattern (expr_iter))

def parse_literal_token (tok):
    if tok[0] is token.NUMBER or tok[0] is token.STRING or \
       (tok[0] is token.NAME and tok[1] in ['True', 'False', 'None']):
        return PLit (tok)

def parse_listlike (expr_iter, parser, ending, allowsplat=True):
    parsed, started = list(), False
    while True:
        tok = expr_iter.peek ()
        if isop (ending, tok):
            expr_iter.next ()
            return parsed
        elif started:
            if not isop (',', tok):
                fail ('expected comma or "%s" in sequence' % ending, tok)
            expr_iter.next ()
        if allowsplat and isop ('*', expr_iter.peek ()):
            splat_tok = expr_iter.next ()
            name_tok = expr_iter.next ()
            if name_tok[0] is not token.NAME:
                fail ('expected variable to bind splat', name_tok)
            parsed.append (PSplat (name_tok[1]))
            allowsplat = False
        else:
            parsed.append (parser (expr_iter))
        if not started: started = True

### Parsing to Python Line objects and Back ###

def lines (reader):
    logical = list ()
    for tok in tokenize.generate_tokens (reader):
        if tok[0] is token.DEDENT:
            if logical:
                for line in __break_tokens (logical):
                    yield line
            yield (END, [])
        elif tok[0] not in [ token.INDENT, tokenize.NL, tokenize.COMMENT, token.ENDMARKER ]:
            logical.append (tok)
        if tok[0] is token.NEWLINE:
            block_index = __block_operator_index (logical)
            for line in __break_tokens (logical):
                yield line
            logical = list ()

def __break_tokens (logical):
    block_index = __block_operator_index (logical)
    if block_index is None:
        return [(LINE, logical)]
    elif block_index == len(logical) - 2:
        return [(BLOCK, logical)]
    else:
        return [(BLOCK, logical[:block_index+1] + [logical[-1]]),
                (LINE, logical[block_index+1:]),
                (END, [])]

def __block_operator_index (tokens):
    "Returns the index of the start block operator in the token list"
    nesting, index, ignore1 = 0, 0, False
    while index < len (tokens):
        if tokens[index][0] == token.OP:
            if tokens[index][1] == ':' and ignore1:
                ignore1 = False
            elif tokens[index][1] == '[' or tokens[index][1] == '{':
                nesting += 1
            elif tokens[index][1] == ']' or tokens[index][1] == '}':
                nesting -= 1
            elif tokens[index][1] == ':' and nesting == 0:
                return index
        elif tokens[index][0] == token.NAME and tokens[index][1] == 'lambda':
            ignore1 = True
        index += 1

def unlines (lines, spaces=4):
    indent = 0
    for (linetype, tokens) in lines:
        if tokens:
            tokens = [ token[:2] for token in tokens ] # Work around http://bugs.python.org/issue12691
            yield ' ' * indent * spaces + tokenize.untokenize (tokens)
        if linetype in INDENTS:
            indent += 1
        elif linetype is END:
            indent -= 1

def extract_block (lines, indent=1):
    while indent > 0:
        line = lines.next ()
        if line[0] in INDENTS:
            indent += 1
        elif line[0] is END:
            indent -= 1
        yield line

### Token Processing ###

def iskeyword (keyword, tok):
    return tok is not None and tok[0] is token.NAME and tok[1] == keyword

def isop (op, tok):
    return tok is not None and tok[0] is token.OP and tok[1] == op

class PeekableIterator(object):
    def __init__(self, iterator):
        self.iterator = iter(iterator)
        self.peeked = None
    
    def __iter__(self):
        return self
    
    def next(self):
        if self.peeked is None:
            return self.iterator.next()
        else:
            previously_peeked = self.peeked[0]
            self.peeked = None
            return previously_peeked
    
    def peek(self):
        if self.peeked is None:
            try:
                self.peeked = (self.iterator.next(),)
            except StopIteration:
                return None
        return self.peeked[0]
    

### Error Handling ###

class ParseError (Exception):
    "Signals an error occured during parsing"
    def __init__(self, message, tok=None, filename=None):
        super (ParseError, self).__init__ (message)
        self.filename = filename
        if tok is not None:
            (self.lineno, self.offset) = tok[2]
            self.token = tok
        else:
            self.lineno = self.offset = self.token = None

def fail (reason, tok=None):
    raise ParseError(reason, tok)

def fail_on_keywords (tokens, *keywords):
    "Fails upon encountering any of the specified keywords in the tokens"
    for tok in tokens:
        for keyword in keywords:
            if iskeyword (keyword, tok):
                fail ('illegal use of keyword "%s"' % keyword, tok)

def print_failure (err, dev=sys.stderr):
    if err.token is not None:
        if err.filename and err.lineno:
            print >>dev, '  File "%s", line %d' % (err.filename, err.lineno)
        elif err.lineno:
            print >>dev, '  Line %d' % err.lineno
        if err.token and len (err.token) > 4:
            print >>dev, err.token[4]
            print >>dev, ' ' * err.offset + '^'
    print >>dev, '%s: %s' % (err.__class__.__name__, err.message)

if __name__ == '__main__':
    from cStringIO import StringIO
    
    if len (sys.argv) != 2:
        print >>sys.stderr, 'usage: python %s filename' % sys.argv[0]
        sys.exit (1)
    
    def unparse (lines):
        for line in lines:
            if line[0] is MATCH:
                yield (BLOCK, [(token.NAME, 'match')] + line[1] + [(token.NAME, 'with'), (token.OP, ':'), (token.NEWLINE, '\n')])
            elif line[0] is CASE:
                predicates = list (tokenize.generate_tokens (StringIO (repr (line[1][0])).readline))[:-1]
                if line[1][1] is None:
                    yield (BLOCK, predicates + [(token.OP, ':'), (token.NEWLINE, '\n')])
                else:
                    yield (BLOCK, predicates + [(token.NAME, 'if')] + line[1][1] + [(token.OP, ':'), (token.NEWLINE, '\n')])
            else:
                yield line
    
    buf = StringIO()
    try:
        for chunk in unlines (unparse (parse_file (sys.argv[1]))):
            buf.write (chunk)
        print buf.getvalue ()
    except ParseError, err:
        print_failure (err)
