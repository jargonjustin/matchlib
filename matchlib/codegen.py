import thread, token
import matchlib.pattern_parser as parser

class CodeGenerator (object):
    def __init__ (self, target=None):
        if target is None: target = list ()
        self.target = target
        self.indent = 0
    
    ### Emitting ###
    
    def emit (self, line):
        self.target.append (line)
    
    def inline (self, lines):
        self.target.extend (lines)
    
    ### Lines ###
    
    def line (self, tokens):
        self.emit ((parser.LINE, tokens + [(token.NEWLINE, '\n')]))
    
    def block (self, tokens):
        self.emit ((parser.BLOCK, tokens + [(token.OP, ':'), (token.NEWLINE, '\n')]))
        self.indent += 1
    
    def dedent (self, amount=1):
        for _ in range(amount):
            self.emit ((parser.END, []))
        self.indent -= amount
    
    ### Statements ###
    
    def Def (self, name, *params):
        self.block (Id ('def') + Call (name, *params))
    
    def Return (self, value):
        self.line (Id ('return') + value)
    
    def Raise (self, *args):
        self.line (Id ('raise') + sepBy (args, Op (', ')))
    
    def If (self, condition):
        self.block (Id('if') + condition)
    
    def ElIf (self, condition):
        self.block (Id('elif') + condition)
    
    def Assign (self, lhs, rhs):
        self.line (lhs + Op('=') + rhs)

### Token Generation ###

# Use a lock for thread-safety
__id_lock = thread.allocate_lock ()
__id_count = 0
def UniqueName ():
    global __id_count, __id_lock
    __id_lock.acquire ()
    __id_count += 1
    ident = '__matchlib_id%d' % __id_count
    __id_lock.release ()
    return ident

def lift (value):
    if isinstance (value, int) or isinstance (value, float):
        return [(token.NUMBER, str (value))]
    elif value is None or value is True or value is False:
        return [(token.NAME, repr (value))]
    elif isinstance (value, str):
        return [(token.STRING, repr (value))]
    else:
        raise TypeError, 'unable to lift %s as a token' % repr (value)

def sepBy (exps, seperator):
    tokens = list ()
    for i, exp in enumerate (exps):
        if i > 0: tokens.extend (seperator)
        tokens.extend (exp) 
    return tokens

def Token (tok): return [tok]

def Op (op): return [(token.OP, op)]
def Id (name): return [(token.NAME, name)]

def Eq (lhs, rhs): return lhs + Op ('==') + rhs
def Gt (lhs, rhs): return lhs + Op ('>') + rhs
def In (lhs, rhs): return lhs + Id ('in') + rhs
def Method (obj, name): return obj + Op ('.') + Id (name)

def Is (lhs, rhs): return lhs + Id ('is') + rhs
def IsNot (lhs, rhs): return lhs + [(token.NAME, 'is'), (token.NAME, 'not')] + rhs

def Index (obj, index): return obj + Op ('[') + index + Op (']')
def Slice (obj, left=None, right=None):
    if left is None: left = []
    if right is None: right = []
    return obj + Op ('[') + left + Op (':') + right + Op (']')

def And (*exps): return sepBy (exps, Id ('and'))
def Tuple (*exps): return Op('(') + sepBy (exps, Op (',')) + Op (',') + Op(')')
def Call (func, *args): return func + Op ('(') + sepBy (args, Op (',')) + Op (')')

