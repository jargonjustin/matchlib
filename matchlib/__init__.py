import matchlib.ast
import matchlib.codec
import matchlib.codegen
import matchlib.compiler
import matchlib.debug
import matchlib.extractors
import matchlib.pattern_parser

__author__ = 'Justin Bonnar <justinbonnar@gmail.com>'
__version__ = '0.1'
