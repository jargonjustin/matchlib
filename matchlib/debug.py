import tokenize, re
import parser as pyparser

import matchlib.pattern_parser as parser

### Anotations ###
# Adds annotation comments to the lines containing debugging information
# sufficient to reconstruct the original file and line information for
# debugging in tracebacks

def annotate (lines, filename=None):
    for line in lines:
        if line[1]:
            (generated, startLine, endLine) = line_info (line[1])
            if startLine is not None:
                comment = 'line %s' % startLine
                if endLine != startLine:
                    comment += '-%s' % endLine
                if filename is not None:
                    comment += ' (%s)' % filename
                if generated:
                    comment += ' (generated)'
                yield (line[0], comment_line (line[1], comment))
                continue
        yield line

def line_info (tokens):
    generated, startLine, endLine = False, None, None
    for token in tokens:
        if len (token) > 2:
            if startLine is None or startLine > token[2][0]:
                startLine = token[2][0]
            if endLine is None or endLine < token[2][0]:
                endLine = token[2][0]
        else:
            generated = True
    return (generated, startLine, endLine)

def comment_line (tokens, comment):
    comment = (tokenize.COMMENT, ' # %s' % comment)
    tokens = list (tokens)
    index = 0
    while tokens[index - 1][0] in [ tokenize.NEWLINE, tokenize.NL ]:
        index -= 1
    if index is 0:
        tokens.append (comment)
    else:
        tokens.insert (index, comment)
    return tokens

### Traceback Rewritting ###
# Rewrites an exception traceback raised at runtime, using debugging annotations
# from compiled .mpy files

__debugging_comment_re = re.compile (r'''
    ^(.*)\#\s+                  # source code and comment begin
    line\s+(\d+)(-\d+)?         # original line(s)
    (:?\s+\(([^)]+)\))?         # original filename
    (:?\s+(\(generated\)))?     # indicates partially generated source
''', re.VERBOSE)
def rewrite_frame (frame):
    (filename, lineno, funcname, text) = frame
    if funcname.startswith ('__matchlib_id'):
        return None
    else:
        annotation = __debugging_comment_re.match (text)
        if annotation is not None:
            return (annotation.group (5), int (annotation.group (2)), funcname, annotation.group (1))
        else:
            return frame

def rewrite_extracted_tb (frames):
    return filter(None, map(rewrite_frame, frames))

def rewrite_syntactic_frame (frame):
    (filename, lineno, offset, text) = frame
    annotation = __debugging_comment_re.match (text)
    if annotation is not None:
        return (annotation.group (5), int (annotation.group (2)), offset, annotation.group (1))
    else:
        return frame

def excepthook(exctype, exc, tb):
    pass

### Static Syntax Checking ###
# Checks to ensure the compiled output is syntactically valid

def check_syntax (source):
    try:
        pyparser.suite (source)
    except IndentationError, err:
        raise IndentationError (err.msg, rewrite_syntactic_frame (err.args[1]))
    except SyntaxError, err:
        raise SyntaxError (err.msg, rewrite_syntactic_frame (err.args[1]))

