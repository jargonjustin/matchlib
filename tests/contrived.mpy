from matchlib.extractors import extractor

def test (value):
    @extractor
    def seq(obj):
        try:
            return tuple ([ elm for elm in iter (obj) ])
        except TypeError:
            return None
    match value with:
        case seq (1, 1, 2, 3, 5, *rest):
            return 'fibonacci'
        case (tag, seq(*_) as details) if isinstance (tag, str):
            return details
        case [head, *tail], (head, *tail):
            return (head, tail)
        case { 'string':  val } if isinstance (val, str):
            return val
        case _:
            return 'no match'

if __name__ == '__main__':
    def fib (max=None):
        a, b = 0, 1
        while max is None or max > a:
            a, b = b, a + b
            yield a
    
    print test ([1, 2, 3])
    print test ({ 'string': 'foo' })
    print test (tuple (fib (20)))
    print test (('a tag', [1, 2, 3]))
    print test ('something else')
