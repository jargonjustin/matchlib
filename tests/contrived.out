(1, [2, 3])
foo
fibonacci
[1, 2, 3]
no match
